content/post/hercules.v4.md

```diff
@@ -0,0 +1,295 @@
+does not happen on GitHub but rather internally at Google and the cumulative
+merges are performed from time to time. It can also be seen that all the top 20
+contributors are from Google.
+
+#### Developers interaction
+
+{{% caption src="/post/hercules.v4/churn_matrix.png" %}}
+Mutual edits in [tensorflow/tensorflow](https://github.com/tensorflow/tensorflow).
+{{% /caption %}}
+
+This analysis shows what are the proportions of lines added by a developer (rows)
+and overwritten by other developers (columns). E.g. the image above shows
+that the Gardener bot overwrote nearly everybody. It also shows that
+[François Chollet](https://github.com/fchollet) worked on the same code as
+[Yifei Feng](https://github.com/yifeif) and Hanna Revinskaya.
+How to reproduce the image:
+
+```
+hercules --burndown --burndown-people --pb https://github.com/tensorflow/tensorflow | tee tensorflow.pb | labours.py -m churn_matrix
+```
+
+#### Files, developers, functions and classes in 3D
+
+If haven't read [Your Code as a Crime Scene](https://pragprog.com/book/atcrime/your-code-as-a-crime-scene)
+by Adam Tornhill, stop reading this post, allocate a few hours and look it through - the book is awesome.
+One of the ideas from the book is to investigate which files are _coupled_ - often appear together.
+The corresponding analysis in Hercules goes a bit further and puts all the files
+into the 3D space with the property that the distance between coupled files is small.
+
+{{% caption src="/post/hercules.v4/couples.png" %}}
+Files coupling in [tensorflow/tensorflow](https://github.com/tensorflow/tensorflow). [Click here for an interactive experience.](http://projector.tensorflow.org/?config=http://vmarkovtsev.github.io/projector/tf_files.json)
+{{% /caption %}}
+
+Reproduce:
+```
+hercules --couples --pb https://github.com/tensorflow/tensorflow | labours.py -m couples -o tensorflow
+```
+
+We can repeat the same trick with the developers: they are couples if they commit
+to the same files. [Interactive example on Tensorflow developers.](http://projector.tensorflow.org/?config=http://vmarkovtsev.github.io/projector/tf_people.json)
+
+Finally, the trick works once again with structural units in the source code. We parse
+each file with [Babelfish](https://doc.bblf.sh) and extract anything we wish: functions,
+classes, etc. Then we apply the same coupling analysis to them: we know which are changed
+in each commit.
+
+By default the functions are extracted.
+```
+hercules --shotness --pb https://github.com/tensorflow/tensorflow | labours.py -m shotness -o tensorflow
+```
+
+#### Comment sentiment
+
+This is a proof-of-concept for running a Tensorflow model over the source code.
+We take [BiDiSentiment](https://github.com/vmarkovtsev/BiDiSentiment) and apply it to comments.
+That model is general purpose and the result is often weird, but it works.
```

Comment:
`That model is general purpose and` -> `That model is a general purpose one and`?